### DISCUSSION TITLE
Running a compute-intensive background task

### DISCUSSION THREAD

**Discussion by jmgilman** [fastapi/fastapi#4554] [Questions] [answered]
Hi,

Trying to wrap my head around creating a background task in my FastAPI app. I have a compute-intensive process that needs to run when the application starts but I don't want it to interfere with receiving requests. I'm still a little new to `asyncio` in general, but I'm not entire unfamiliar with async concepts. My initial idea was to create a task that runs the compute-intensive code:

```
asyncio.create_task(my_compute_intensive_task)
```

However, what I noticed is that when my application starts all incoming requests hang until this compute-intensive task finishes. In my mind, this makes sense, because there's no actual I/O bound operation going on that would ever yield, so the task ends up eating all the compute until it's done. 

My question is what is the correct way to handle a compute-intensive task without interrupting requests? 

My real-life scenario has some additional complexity - mainly in that *some* requests are dependent on the compute-intensive task being finished. For these exceptions, I assumed I would use a lock of sorts - but without knowing the right way to do things it's hard for me to tell what the correct solution is.

Any help is appreciated!

**Comment by upeshjindal** [comment]
For very heavy tasks, FastAPI recommends to handle them through dedicated solutions like Celery so that they run outside the process and hence don't block the FastAPI requests.

**Comment by jmgilman** [comment]
That's understandable, but I wouldn't classify this task as very heavy. It takes approx. 10 seconds to complete, but the entire FastAPI app is running in a serverless environment, meaning cold startups are likely and it's important to minimize startup time. The way I have it implemented now it takes a little over 10 seconds before the app actually starts responding to requests (due to the compute-intensive task). With the way GCR (my serverless provider) works, this means that it's possible for a request to see an unusually slow response due to this startup time. 

Normally, I would just run this code once and cache the result, but it's too large for a cache which is the reason why I need it to run once at container startup. Something like celery seems overkill for this as I really just need this 10 second task to run and not interfere with `uvcorn's` ability to respond to requests. 

Hope that makes more sense.

**Reply by upeshjindal** [reply]
I see. If this needs to run once at container startup and I assume this means running it at app startup, then we can put it in the startup event of FastAPI and not necessarily as a background task because it appears that the app depends on it.

Another option is to run it as another command in a script just before you execute uvicorn/gunicorn.

I hope I'm able to connect to your explanation.

**Reply by jmgilman** [reply]
Yes, the app startup method is where I currently have the code at. That's where I'm using `asyncio` to create a task which runs the code. The problem is that this setup bogs the application down until it finishes and results in the application not responding to connections for the first 10 seconds.

My goal is to run this startup code in such a way as to not impose upon incoming connections during the first 10 seconds. It's important to note that GCR spawns containers due to increases in traffic - meaning there is almost always an end-user waiting during container startup - so I would like to parallelize the startup method if possible. I'm just not sure how to do this in the context of FastAPI and `asyncio`.

**Reply by SDAravind** [reply]
@jmgilman - There are best practices mentioned at [FastAPI best practices](https://github.com/zhanymkanov/fastapi-best-practices#7-dont-make-your-routes-async-if-you-have-only-blocking-io-operations) using threads. Code example are pretty neat.

**Comment by jgould22** [comment, marked as answer]
10s is an eternity for an event loop.

What if you spawned a process outside of the loop that will do your computing and then stores the result to a file.

The endpoints that need it can load the file the first time it is needed and then cache the file for future calls.

**Reply by upeshphable** [reply]
My take is also the same on this issue. I think it should be an external task.

**Comment by pawamoy** [comment]
If you have control over the CPU-intensive code, you can maybe force context switches by dispatching `asyncio.sleep(0.05)`  lines in the code.
